measurements/measure.cpp

```diff
@@ -0,0 +1,335 @@
+
+public:
+    XRcuPtr(size_t vec_size)
+        : v(asp_traits::make_shared<std::vector<int>>(vec_size,
+                                                      default_value)) {}
+
+    int read_one(unsigned index) const {
+        asp_traits::shared_ptr<const std::vector<int>> local_copy = v.read();
+        assert(index < local_copy->size());
+        return (*local_copy)[index];
+    }
+    int read_all() const { // sum
+        asp_traits::shared_ptr<const std::vector<int>> local_copy = v.read();
+        return std::accumulate(local_copy->begin(), local_copy->end(), 0);
+    }
+
+    void update_all(int value) {
+        v.copy_update([=](std::vector<int>* copy) {
+            for (auto& e : *copy) {
+                e = value;
+            }
+        });
+    }
+};
+
+class XStdMutex {
+    std::vector<int> v;
+    const int default_value = 1;
+    mutable std::mutex m;
+
+public:
+    XStdMutex(size_t vec_size) : v(vec_size, default_value) {}
+
+    int read_one(unsigned index) const {
+        std::lock_guard<std::mutex> lock{m};
+        assert(index < v.size());
+        return v[index];
+    }
+    int read_all() const { // sum
+        std::lock_guard<std::mutex> lock{m};
+        return std::accumulate(v.begin(), v.end(), 0);
+    }
+
+    void update_all(int value) {
+        std::lock_guard<std::mutex> lock{m};
+        for (auto& e : v) {
+            e = value;
+        }
+    }
+};
+
+class XTbbQueuingRwMutex {
+    std::vector<int> v;
+    const int default_value = 1;
+    mutable tbb::queuing_rw_mutex m;
+
+public:
+    XTbbQueuingRwMutex(size_t vec_size) : v(vec_size, default_value) {}
+
+    int read_one(unsigned index) const {
+        tbb::queuing_rw_mutex::scoped_lock lock{m, false}; // read lock
+        assert(index < v.size());
+        return v[index];
+    }
+    int read_all() const {                                 // sum
+        tbb::queuing_rw_mutex::scoped_lock lock{m, false}; // read lock
+        return std::accumulate(v.begin(), v.end(), 0);
+    }
+
+    void update_all(int value) {
+        tbb::queuing_rw_mutex::scoped_lock lock{m}; // write lock
+        for (auto& e : v) {
+            e = value;
+        }
+    }
+};
+
+class XTbbSpinRwMutex {
+    std::vector<int> v;
+    const int default_value = 1;
+    mutable tbb::spin_rw_mutex m;
+
+public:
+    XTbbSpinRwMutex(size_t vec_size) : v(vec_size, default_value) {}
+
+    int read_one(unsigned index) const {
+        tbb::spin_rw_mutex::scoped_lock lock{m, false}; // read lock
+        assert(index < v.size());
+        return v[index];
+    }
+    int read_all() const {                              // sum
+        tbb::spin_rw_mutex::scoped_lock lock{m, false}; // read lock
+        return std::accumulate(v.begin(), v.end(), 0);
+    }
+
+    void update_all(int value) {
+        tbb::spin_rw_mutex::scoped_lock lock{m}; // write lock
+        for (auto& e : v) {
+            e = value;
+        }
+    }
+};
+
+class XURCU {
+    std::vector<int>* v;
```

Comment:
How about using `std::unique_ptr<>` instead?

Reply:
I am not sure if we could use it, especially `rcu_assign_pointer` seems problematic with the smart ptr